### PR TITLE
More tests to Trainer

### PR DESCRIPTION
**Pull request by sgugger**
While doing see, realized there were some problems with the seed (in particular for HP search) so added a few tests of that too.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -77,6 +77,7 @@ jobs:
                       - v0.3-torch_and_tf-{{ checksum "setup.py" }}
                       - v0.3-{{ checksum "setup.py" }}
             - run: pip install --upgrade pip
+            - run: pip install git+https://github.com/huggingface/nlp
```
Comment (sgugger) [inline .circleci/config.yml:80]: For now, using nlp master branch to test the integration with Trainer (we need some unreleased features). Not sure if we want to keep it this way for the foreseeable future or use the last release later on.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -206,22 +206,29 @@ def __init__(
...
-        self.args = args
         self.data_collator = data_collator if data_collator is not None else default_data_collator
         self.train_dataset = train_dataset
         self.eval_dataset = eval_dataset
         self.model_init = model_init
         self.compute_metrics = compute_metrics
         self.optimizer, self.lr_scheduler = optimizers
+        if model_init is not None and (self.optimizer is not None or self.lr_scheduler is not None):
+            raise RuntimeError(
+                "Passing a `model_init` is incompatible with providing the `optimizers` argument."
```
Comment (sgugger) [inline src/transformers/trainer.py:229]: Raising an error as a defense mechanism: `model_init` means each new training starts from scratch, thus needs a clean optimizer/scheduler.

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -821,20 +833,22 @@ def hyperparameter_search(
...
         if backend == HPSearchBackend.RAY and not is_ray_available():
             raise RuntimeError(
-                " You picked the Ray Tune backend, but it is not installed. Use `pip install 'ray[tune]'`."
+                "You picked the Ray Tune backend, but it is not installed. Use `pip install 'ray[tune]'`."
             )
         self.hp_search_backend = backend
 
+        if self.model_init is None:
+            raise RuntimeError(
+                "To use hyperparameter search, you need to pass your model through a model_init function."
```
Comment (sgugger) [inline src/transformers/trainer.py:845]: Defense mechanism: you can't used HP search if the model is not reinitialized at each training.

File: src/transformers/trainer.py [hunk 4]
```diff
@@ -206,22 +206,29 @@ def __init__(
         optimizers: Tuple[torch.optim.Optimizer, torch.optim.lr_scheduler.LambdaLR] = (None, None),
         **kwargs,
     ):
+        if args is None:
+            logger.info("No `TrainingArguments` passed, using the current path as `output_dir`.")
+            args = TrainingArguments("tmp_trainer")
+        self.args = args
+        # Seed must be set before instantiating the model when using model
+        set_seed(self.args.seed)
```
Comment (LysandreJik) [inline src/transformers/trainer.py:214]: Good catch!

**Review by LysandreJik — APPROVED** [review]
Great, love the tests!
